### PR TITLE
[dynamo][eval_frame] Set the callback to None earlier for guard eval

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139038
* __->__ #139655

xref - https://fb.workplace.com/groups/1075192433118967/permalink/1536570810314458/

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/csrc/dynamo/eval_frame.c [hunk 1]
```diff
@@ -611,6 +611,12 @@ static PyObject* _custom_eval_frame(
 
   PyObject* backend = get_backend(callback);
 
+
+  // We don't run the current custom_eval_frame behavior for guards.
+  // So we temporarily set the callback to Py_None to drive the correct behavior
+  // in the shim.
+  eval_frame_callback_set(Py_None);
```
Comment (anijain2305) [inline torch/csrc/dynamo/eval_frame.c:618]: This ensure that for torch._dynamo.run mode or `extra_state_cache_limit_hit(extra)` also has callback disabled (condition at L624)

**Review by jansel — APPROVED** [review]

File: torch/csrc/dynamo/eval_frame.c [hunk 2]
```diff
@@ -645,18 +651,15 @@ static PyObject* _custom_eval_frame(
     PyCodeObject* cached_code = (PyCodeObject*)maybe_cached_code;
     // used cached version
     DEBUG_TRACE("cache hit %s", get_frame_name(frame));
+    // Re-enable custom behavior
+    eval_frame_callback_set(callback);
```
Comment (williamwen42) [inline torch/csrc/dynamo/eval_frame.c:655]: Assuming we don't need to also do this in the `return NULL` line (L637) because we'd be crashing there.

**Review by williamwen42 — APPROVED** [review]

File: torch/csrc/dynamo/eval_frame.c (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/eval_frame.c:618]: Yes.

File: torch/csrc/dynamo/eval_frame.c (hunk 2 above)
Comment (anijain2305) [reply torch/csrc/dynamo/eval_frame.c:655]: Yes.

**Pull request closed without merging**